core/src/contracts/stablex_contract.rs

```diff
@@ -437,4 +516,49 @@ pub mod tests {
             (expected_prices, expected_token_ids)
         );
     }
+
+    #[test]
+    fn incremental_binary_search() {
+        struct MockBlockchain {
```

Comment:
Nice! One pro-tip is we have a `mockall` crate, which can automatically generate these mock implementations for you.